moveit_core/collision_detection/include/moveit/collision_detection/collision_common.h

```diff
@@ -217,6 +220,185 @@ struct CollisionRequest
+    : detailed(detailed)
+    , global(global)
+    , group_name(group_name)
+    , active_components_only(NULL)
+    , acm(&acm)
+    , distance_threshold(distance_threshold)
+    , verbose(false)
+    , gradient(false)
+  {
+  }
+
+  virtual ~DistanceRequest()
+  {
+  }
+
+  /// Compute \e active_components_only_ based on \e req_
+  void enableGroup(const robot_model::RobotModelConstPtr& kmodel)
+  {
+    if (kmodel->hasJointModelGroup(group_name))
+      active_components_only = &kmodel->getJointModelGroup(group_name)->getUpdatedLinkModelsWithGeometrySet();
+    else
+      active_components_only = NULL;
+  }
+
+  /// Indicate if detailed information should be calculated
+  bool detailed;
+
+  /// Indicate if the global minimum should be found
+  bool global;
+
+  std::string group_name;
+
+  /// The set of active components to check
+  const std::set<const robot_model::LinkModel*>* active_components_only;
+
+  /// The allowed collision matrix used to filter checks
+  const AllowedCollisionMatrix* acm;
+
+  /// A distance threshold to reduce number of queries
+  double distance_threshold;
+
+  /// Log debug information
+  bool verbose;
+
+  /// Indicate if gradient should be calculated between each object. This the vector defined by the line connecting the
+  /// closest points on the two objects.
+  bool gradient;
+};
+
+struct DistanceResultsData
+{
+  DistanceResultsData()
+  {
+    clear();
+  }
+
+  /// The minimum distance between two objects. If two objects are in collision, min_distance <= 0.
+  double min_distance;
+
+  /// The nearest points
+  Eigen::Vector3d nearest_points[2];
+
+  /// The object link names
+  std::string link_name[2];
+
+  /// The gradient
+  Eigen::Vector3d gradient;
+
+  /// Indicates if gradient was calculated.
+  bool hasGradient;
+
+  /// Indicates if nearest points were found.
+  bool hasNearestPoints;
+
+  /// Clear structure data
+  void clear()
+  {
+    min_distance = std::numeric_limits<double>::max();
+    nearest_points[0].setZero();
+    nearest_points[1].setZero();
+    link_name[0] = "";
+    link_name[1] = "";
+    gradient.setZero();
+    hasGradient = false;
+    hasNearestPoints = false;
+  }
+
+  /// Update structure data given DistanceResultsData object
+  void update(const DistanceResultsData& results)
+  {
+    min_distance = results.min_distance;
+    nearest_points[0] = results.nearest_points[0];
+    nearest_points[1] = results.nearest_points[1];
+    link_name[0] = results.link_name[0];
+    link_name[1] = results.link_name[1];
+    gradient = results.gradient;
+    hasGradient = results.hasGradient;
+    hasNearestPoints = results.hasNearestPoints;
+  }
+};
+
+typedef std::map<std::string, DistanceResultsData> DistanceMap;
+
+struct DistanceResult
+{
+  DistanceResult() : collision(false)
+  {
+  }
+  virtual ~DistanceResult()
+  {
+  }
+
+  /// Indicates if two objects were in collision
+  bool collision;
+
+  /// The results for the two objects with the minimum distance
```

Comment:
I think this needs more documentation. What's the difference between `minimum_distance` and `distance`?
What exactly is stored in the `DistanceMap`?

Reply:
The DistanceMap stores distance results information for every link in active components list.